cassandra/connection.py

```diff
@@ -732,3 +743,96 @@ def deliver(self, timeout=None):
             raise OperationTimedOut()
         else:
             return self.responses
+
+
+class HeartbeatFuture(object):
+    def __init__(self, connection, owner):
+        self._exception = None
+        self._event = Event()
+        self.connection = connection
+        self.owner = owner
+        log.debug("Sending options message heartbeat on idle connection %s %s",
+                  id(connection), connection.host)
+        with connection.lock:
+            if connection.in_flight < connection.max_request_id:
+                connection.in_flight += 1
+                connection.send_msg(OptionsMessage(), connection.get_request_id(), self._options_callback)
+            else:
+                self._exception = Exception("Failed to send heartbeat because connection 'in_flight' exceeds threshold")
```

Comment:
I think this may be problematic for the v1/v2 protocol under load, because connections can get maxed out pretty easily.  Here are a few alternatives:
* Wait for in_flight to drop in a bounded loop with sleeps.  This is kind of expensive, but should happen pretty quickly with a healthy connection.
* Skip the heartbeat for this round.  If in_flight is maxed out because the connection is healthy, this is fine.  If it's maxed out because none of the operations are completing, then, well ... skipping the round sucks.  So I guess this option isn't great.
* Always reserve a spare request ID for heartbeats (by lowering the max for other connection users).  This is probably the best option, but requires a bit more work.